### PR TITLE
Merge "normals" into master

### PR DESCRIPTION
**Pull request by JasperDre**
Add vertex normal and light rendering.

### CODE REVIEW

File: source/Subsystem/Renderer/RenderContext.cpp [hunk 1]
```diff
@@ -82,21 +89,27 @@ void RenderContext::CreateBuffers(std::vector<Model>& aModels)
 		glEnableVertexAttribArray(1);
 		glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, sizeof(Vertex), (GLvoid*)offsetof(Vertex, myTextureCoordinates));
 
+		glEnableVertexAttribArray(2);
+		glVertexAttribPointer(2, 3, GL_FLOAT, GL_FALSE, sizeof(Vertex), (GLvoid*)offsetof(Vertex, myNormal));
+
+		glEnableVertexAttribArray(3);
+		glVertexAttribPointer(3, 3, GL_FLOAT, GL_FALSE, sizeof(Vertex), (GLvoid*)offsetof(Vertex, myTangent));
+
```
Comment (florianschut) [inline source/Subsystem/Renderer/RenderContext.cpp:97]: C-Style casts?!!?!?!?!?!

Comment (florianschut) [reply source/Subsystem/Renderer/RenderContext.cpp:97]: Tangent currently is unused in the shaders but will be necesarry for normal mapping!

File: source/Subsystem/Renderer/AssetLoader.cpp [hunk 2]
```diff
@@ -200,13 +224,49 @@ Model* AssetLoader::LoadOBJ(const std::string& aPath, const std::string& aFilena
...
+        Tangent.z = f * (DeltaV2 * Edge1.z - DeltaV1 * Edge2.z);
+
+        Bitangent.x = f * (-DeltaU2 * Edge1.x - DeltaU1 * Edge2.x);
+        Bitangent.y = f * (-DeltaU2 * Edge1.y - DeltaU1 * Edge2.y);
+        Bitangent.z = f * (-DeltaU2 * Edge1.z - DeltaU1 * Edge2.z);
+
+        v0.myTangent += Tangent;
+        v1.myTangent += Tangent;
+        v2.myTangent += Tangent;
+    }
```
Comment (florianschut) [inline source/Subsystem/Renderer/AssetLoader.cpp:256]: Tangents is currently unused but will be needed for normal mapping

File: source/Subsystem/Renderer/AssetLoader.cpp [hunk 3]
```diff
@@ -200,13 +224,49 @@ Model* AssetLoader::LoadOBJ(const std::string& aPath, const std::string& aFilena
...
+
+        glm::vec3 Tangent, Bitangent;
+
+        Tangent.x = f * (DeltaV2 * Edge1.x - DeltaV1 * Edge2.x);
+        Tangent.y = f * (DeltaV2 * Edge1.y - DeltaV1 * Edge2.y);
+        Tangent.z = f * (DeltaV2 * Edge1.z - DeltaV1 * Edge2.z);
+
+        Bitangent.x = f * (-DeltaU2 * Edge1.x - DeltaU1 * Edge2.x);
+        Bitangent.y = f * (-DeltaU2 * Edge1.y - DeltaU1 * Edge2.y);
+        Bitangent.z = f * (-DeltaU2 * Edge1.z - DeltaU1 * Edge2.z);
```
Comment (florianschut) [inline source/Subsystem/Renderer/AssetLoader.cpp:251]: What does this do? seems the result is going out of scope without being used

File: source/Subsystem/Renderer/RenderContext.h [hunk 4]
```diff
@@ -14,13 +15,15 @@ class RenderContext
...
 	void Destroy(const std::vector<Model>& aModels);
 
 private:
 	void PrintDebugInfo();
 	void CreateCamera();
+	void CreateLight();
 
 private:
 	Camera* myCamera;
+	Light* myLight;
```
Comment (florianschut) [inline source/Subsystem/Renderer/RenderContext.h:28]: This hurts my soul! Could be fixed later tho

**Review by florianschut — CHANGES_REQUESTED** [review]
Bunch of comments that should be addressed xoxo

File: source/Subsystem/Renderer/RenderContext.cpp (hunk 1 above)
Comment (JasperDre) [reply source/Subsystem/Renderer/RenderContext.cpp:97]: True, we can actually calculate and pass the tangents and bitangents without much work.

File: source/Subsystem/Renderer/RenderContext.h (hunk 4 above)
Comment (JasperDre) [reply source/Subsystem/Renderer/RenderContext.h:28]: What would you suggest? Right now I just made a raw pointer to test lighting.

File: source/Subsystem/Renderer/AssetLoader.cpp (hunk 3 above)
Comment (florianschut) [reply source/Subsystem/Renderer/AssetLoader.cpp:251]: Please explain ;)

File: source/Subsystem/Renderer/RenderContext.h (hunk 4 above)
Comment (florianschut) [reply source/Subsystem/Renderer/RenderContext.h:28]: I was referring to the single light stored in the render context... Should be handled by world or wherever scenes are stored. Should be fine for now till we expand on the lighting system!

File: Data/Shaders/NormalVertexShader.glsl [hunk 5]
```diff
@@ -0,0 +1,35 @@
+// VERTEX SHADER
+#version 330 core
+
+layout(location = 0) in vec3 inVertexPositionModelSpace;
+layout(location = 1) in vec2 inVertexUVs;
+layout(location = 2) in vec3 inNormal;
+layout(location = 3) in vec3 inTangent;
+
+out LightData {
+	vec3 objectColor;
```
Comment (florianschut) [inline Data/Shaders/NormalVertexShader.glsl:10]: This is unrelated to light data since it reflects the color of objects in the scene and not a light and therefore should probably not be in LightData

Comment (florianschut) [reply Data/Shaders/NormalVertexShader.glsl:10]: Also consider if you actually need this attribute I haven't seen this used before (only sparsely seen vertex color used)

**Review by florianschut — CHANGES_REQUESTED** [review]
Love the new naming in the shaders!
Found one questionable attribute in LightData

File: source/Subsystem/Renderer/RenderContext.h (hunk 4 above)
Comment (JasperDre) [reply source/Subsystem/Renderer/RenderContext.h:28]: Yes, I agree. We'll probably do that soon.

Comment (florianschut) [reply source/Subsystem/Renderer/RenderContext.h:28]: @JasperDre call me UwU

File: Data/Shaders/NormalFragmentShader.glsl [hunk 6]
```diff
@@ -24,7 +23,7 @@ void main()
 	vec3 lightDirection = normalize(inLightData.lightPosition - vec3(1.0, 1.0, 1.0));
 	float diffuseFactor = max(dot(normalizedNormal, lightDirection), 0.0);
 	vec3 diffuse = diffuseFactor * inLightData.lightColor;
-	vec3 result = (ambient + diffuse) * inLightData.objectColor;
+	vec3 result = (ambient + diffuse);
```
Comment (florianschut) [inline Data/Shaders/NormalFragmentShader.glsl:26]: Nice brackets, will remove in my next commit

**Review by florianschut — APPROVED** [review]
Looks good 💯

File: Data/Shaders/NormalFragmentShader.glsl [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+out vec3 color;
+
+uniform sampler2D textureSampler;
+
+void main()
+{
+	float ambientStrength = 0.1;
+    vec3 ambient = ambientStrength * lightColorOut;
+	vec3 norm = normalize(Normal);
+	vec3 lightDir = normalize(lightPosOut - vec3(1.0, 1.0, 1.0));
```
[the remark this answers is not in the document]
Comment (florianschut) [reply Data/Shaders/NormalFragmentShader.glsl]: Should be fixed in latest [https://github.com/Pyrohax/Simpleton/pull/16/commits/5796fa15c52983de837efc30242a5b885ad1a7a8](commit) @JasperDre could you review? :)

**Comment by florianschut** [comment]
Should remove or do something with Bitangent. Can be merged afterwards in my opinion!
